pkg/controller/kubedirectorcluster/cluster.go

```diff
@@ -57,6 +57,65 @@ func (r *ReconcileKubeDirectorCluster) syncCluster(
 		cr.Status = &kdv1.KubeDirectorClusterStatus{}
 		cr.Status.Roles = make([]kdv1.RoleStatus, 0)
 	}
+
+	// Check, if some statefulset is in the middle of upgrade/rollback process.
+	// If so, the current cluster state still can be in `configured` state, change it to `updating`
+	upgradeIsActive := false
+	for _, rs := range cr.Status.Roles {
+		if rs.RoleUpgradeStatus == kdv1.RoleRollingBack || rs.RoleUpgradeStatus == kdv1.RoleUpgrading {
+			upgradeIsActive = true
+			cr.Status.State = string(clusterUpdating)
+			break
+		}
+	}
+
+	// If all statefulsets completed their upgrade/rollback processes
+	// erase the cluster upgradeInfo object
+	if cr.Status.UpgradeInfo != nil && !upgradeIsActive {
+		// If there wasn't rollback
+		if cr.Spec.AppID != cr.Status.UpgradeInfo.PrevApp {
```

Comment:
Hmm, app reference management. I think we need to make sure we have a reference on both apps as long as an upgrade or rollback is active. Assuming we are doing the change to prevApp on rollback that I mentioned below, then that means:
* On creating upgradeInfo (starting upgrade), add the reference to the app in cr.Spec.AppID.
* On deleting upgradeInfo (ending upgrade or rollback), remove the reference to the app in upgradeInfo.PrevApp.

Reply:
Fixed